Disallow zero sized types

fselmo opened the pull request:
### What was wrong?

- Solidity doesn't support empty structs / tuples; we should not support them in `eth-abi`.

### How was it fixed?

- Validate against zero-sized tuples and provide a good message for these errors.
- Add tests and update current tests for empty tuples.

### Todo:
- [x] Clean up commit history
- [x] Add entry to the [release notes](https://github.com/ethereum/eth-abi/blob/master/newsfragments/README.md)

#### Cute Animal Picture

The Sunda Colugo

![Put a link to a cute animal picture inside the parenthesis-->](https://cdn2.wanderlust.co.uk/media/1027/cropped-shutterstock_170799332.jpg?anchor=center&mode=crop&width=1920&height=0&format=auto&quality=90&rnd=132204600710000000)

fselmo committed:
bugfix: validate against zero-sized tuple types.

- Zero-sized tuple types are not supported by Solidity. We should reflect that in this library.
- Add tests for this and properly set up reversibility tests to test for minimum_size=1 for tuple encoding / decoding tests.

fselmo committed:
Add more encode tests around zero-sized tuple validation

reedsa reviewed (approved):
thanks for the walkthrough, lgtm!

fselmo committed:
Keep zero-sized tuple validation in one spot only.

fselmo committed:
Refactor for zero-sized tuple validation:

- Keep param validation for ``encode()`` and ``decode()`` where they are but the type string validation didn't make a whole lot of sense there. There's a grammar.parse that takes place trying to identify type strings and we give a very general error message when they can't be parsed. Adding some logic here when we know what the parse error is seems like a better, more central spot to take care of better messaging around these types of errors.
- Along the way, add a bit more generalized logic around ``is_encodable_type()`` and ``is_encodable()``. We don't need to sift through different error types when we are looking for a ``bool`` answer. These methods ideally shouldn't throw any errors and adding exception classes to it is just a chore. Catch any exception and return ``False``, otherwise, when there are no errors, return ``True``. The exception to this is when multiple entries are found, that seems like a good case to throw.

reedsa reviewed (approved):
Looks great, thanks for taking the extra time on this one.

fselmo committed:
Newsfragment for #212
